File: qeli-shared/QeliShared/Geo/ProtoReader.cs
using System.Buffers.Binary;

namespace Qeli.Shared.Geo;

/// <summary>Minimal protobuf wire reader for v2ray geosite/geoip .dat files.</summary>
internal ref struct ProtoReader
{
    private ReadOnlySpan<byte> _data;
    private int _pos;

    public ProtoReader(ReadOnlySpan<byte> data)
    {
        _data = data;
        _pos = 0;
    }

    public bool HasMore => _pos < _data.Length;

    public bool TryReadTag(out int fieldNumber, out int wireType)
    {
        fieldNumber = 0;
        wireType = 0;
        if (!HasMore) return false;
        ulong key = ReadVarint();
        fieldNumber = (int)(key >> 3);
        wireType = (int)(key & 0x7);
        return true;
    }

    public ulong ReadVarint()
    {
        ulong result = 0;
        int shift = 0;
        while (_pos < _data.Length)
        {
            byte b = _data[_pos++];
            result |= (ulong)(b & 0x7F) << shift;
            if ((b & 0x80) == 0) return result;
            shift += 7;
            if (shift > 63) throw new InvalidDataException("varint too long");
        }
        throw new InvalidDataException("truncated varint");
    }

    public ReadOnlySpan<byte> ReadBytes()
    {
        int len = (int)ReadVarint();
        if (len < 0 || _pos + len > _data.Length) throw new InvalidDataException("bad length-delimited");
        var s = _data.Slice(_pos, len);
        _pos += len;
        return s;
    }

    public string ReadString() => System.Text.Encoding.UTF8.GetString(ReadBytes());

    public void Skip(int wireType)
    {
        switch (wireType)
        {
            case 0: ReadVarint(); break;
            case 1: _pos += 8; break;
            case 2: _ = ReadBytes(); break;
            case 5: _pos += 4; break;
            default: throw new InvalidDataException($"unsupported wire type {wireType}");
        }
    }
}

File: qeli-android/app/src/main/kotlin/com/qeli/geo/ProtoReader.kt
package com.qeli.geo

/** Minimal protobuf wire reader for v2ray geosite/geoip .dat files. */
internal class ProtoReader(private val data: ByteArray, private var pos: Int = 0) {
    val hasMore: Boolean get() = pos < data.size

    fun tryReadTag(): Pair<Int, Int>? {
        if (!hasMore) return null
        val key = readVarint()
        return ((key ushr 3).toInt()) to ((key and 7L).toInt())
    }

    fun readVarint(): Long {
        var result = 0L
        var shift = 0
        while (pos < data.size) {
            val b = data[pos++].toInt() and 0xff
            result = result or ((b and 0x7f).toLong() shl shift)
            if (b and 0x80 == 0) return result
            shift += 7
            if (shift > 63) throw IllegalArgumentException("varint too long")
        }
        throw IllegalArgumentException("truncated varint")
    }

    fun readBytes(): ByteArray {
        val len = readVarint().toInt()
        if (len < 0 || pos + len > data.size) throw IllegalArgumentException("bad length-delimited")
        val s = data.copyOfRange(pos, pos + len)
        pos += len
        return s
    }

    fun readString(): String = String(readBytes(), Charsets.UTF_8)

    fun skip(wireType: Int) {
        when (wireType) {
            0 -> readVarint()
            1 -> pos += 8
            2 -> readBytes()
            5 -> pos += 4
            else -> throw IllegalArgumentException("unsupported wire type $wireType")
        }
    }
}
